HPCC-31130 Handle MP accept connection negotiations asynchronously

jakesmith opened the pull request:
And handoff slow/blocking connection negotiations to separate dedicated thread.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith committed:
HPCC-31130 Handle MP accept connection negotiations asynchronously

And handoff slow/blocking connection negotiations to separate
dedicated thread.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
@mckellyln - please review.
It is worth noting, that the code in handleAcceptedSocket is an almost wholesale move of the code that used to be in CMPConnectThread::run()

mckellyln commented on system/mp/mpcomm.cpp:
```diff
@@ -2178,194 +2265,272 @@ void CMPConnectThread::startPort(unsigned short port)
+                    {
+                        constexpr unsigned timeoutMs = 5000;
+                        Owned<ISocket> handledSock = sock.getClear();
+                        if (owner.handleAcceptedSocket(handledSock.getLink(), timeoutMs, false))
```
Why try for 5000 ms before handing off to async thread(s) ?  Why not just always use async thread(s) ?
Maybe almost same thing if we change 5000 ms here to 100 ms or something very short ..

jakesmith replied:
It does always use an async thread - this code is in in CMPConnectionThread::threadmain (a thread in the pool).

What it's doing here (within a thread of the pool), is handing this connection off, to outside the thread pool, to a separate 'slow' handler, so pool doesn't clog up.

jakesmith commented on system/mp/mpcomm.cpp:
```diff
@@ -448,11 +448,85 @@ class CMPConnectThread: public Thread
+        std::vector<Owned<ISocket>> slowClientsSocks;
+        CriticalSection crit;
+        Semaphore sem;
+        bool stopped = true;
```
probably, I'll change it.

jakesmith commented:
@mckellyln - made 'stopped' an atomic, and see reply re. other question.

mckellyln commented:
I've done some testing and this PR looks good, but I can still get some bad behavior when a rogue client connects to Thor and sends in various amounts of garbage data.  I have created https://track.hpccsystems.com/browse/HPCC-31251 for additional work on this.

mckellyln reviewed (approved):
I approve this PR.

ghalliday commented:
@jakesmith please can you add a summary of the solution to the jira.

jakesmith commented:
> @jakesmith please can you add a summary of the solution to the jira.

Added. Also noticed a problem with the way the optional config. option was being picked up in k8s. Addressed in new commit.
@ghalliday

ghalliday commented on system/mp/mpcomm.cpp:
```diff
@@ -463,9 +537,16 @@ class CMPConnectThread: public Thread
             listensock->cancel_accept();
+            if (listen)
+            {
+                if (!threadPool->joinAll(true, 1000*60*5))
```
Worth commenting that stop the listening threads first in case any get added to the stop processor?

ghalliday commented on system/mp/mpcomm.cpp:
```diff
@@ -2178,194 +2265,272 @@ void CMPConnectThread::startPort(unsigned short port)
     running = true;
+    if (acceptThreadPoolSize)
+    {
+        class CFactory : public CInterfaceOf<IThreadFactory>
```
CFactory: The name could be more descriptive - what is it a factory for?

jakesmith replied:
Will rename to CMPConnectThreadFactory.
I do tend to use generic names if inline classes are used and therefore implicitly relevant to the surrounding code.

ghalliday commented on system/mp/mpcomm.cpp:
```diff
@@ -2178,194 +2265,272 @@ void CMPConnectThread::startPort(unsigned short port)
+                    virtual void threadmain() override
+                    {
+                        constexpr unsigned timeoutMs = 5000;
+                        Owned<ISocket> handledSock = sock.getClear();
```
I am curious why you move it to a local variable?  To ensure I assume to assume it is detroyed when the thread finishes.  Probably worth a comment.

jakesmith replied:
it's because this is a reusable thread.
I want to ensure when the thread is past this threadmain, the socket is no longer a linked member of the underlying reusable thread object.

Arguably, there should be a IPooledThread::cleanup() or similar, which is called after the pooled thread has completed an execution cycle.

I'll add a comment.

ghalliday commented on system/mp/mpcomm.cpp:
```diff
@@ -463,9 +537,16 @@ class CMPConnectThread: public Thread
             listensock->cancel_accept();
+            if (listen)
+            {
+                if (!threadPool->joinAll(true, 1000*60*5))
```
threadPool could be null if the pool size is set to 0

jakesmith replied:
yes, will add a check.

ghalliday commented on system/mp/mpcomm.cpp:
```diff
@@ -2178,194 +2265,272 @@ void CMPConnectThread::startPort(unsigned short port)
+                        Owned<ISocket> handledSock = sock.getClear();
+                        if (owner.handleAcceptedSocket(handledSock.getLink(), timeoutMs, false))
+                        {
+                            // handoff to slowClientProcessor, which will retry for standard CONFIRM_TIMEOUT period                            owner.slowClientProcessor.add(sock.getClear());
```
spurious (almost) copy of the next line on the end of the comment.

ghalliday commented on system/mp/mpcomm.cpp:
```diff
@@ -2178,194 +2265,272 @@ void CMPConnectThread::startPort(unsigned short port)
+                    {
+                        constexpr unsigned timeoutMs = 5000;
+                        Owned<ISocket> handledSock = sock.getClear();
+                        if (owner.handleAcceptedSocket(handledSock.getLink(), timeoutMs, false))
```
Not worth changing, but it is counter-intuitive that a function called handleAcceptedSocket() returns true if it fails to handle it.

jakesmith replied:
agree, and pondered this, but really true|false is insufficient. As it stands returning false means it did not timeout, but it doesn't mean it succeeded. Various failure conditions also return false.
So really it should return a state, but I only care about 1 of them (timeout).

ghalliday reviewed (commented):
@jakesmith A couple of problems and a few minor other comments.

jakesmith commented on system/mp/mpcomm.cpp:
```diff
@@ -448,11 +448,85 @@ class CMPConnectThread: public Thread
+                    continue;
+                }
+
+                Linked<ISocket> sock = slowClientsSocks.back();
```
I will put both in crit block.

jakesmith commented on system/mp/mpcomm.cpp:
```diff
@@ -448,11 +448,85 @@ class CMPConnectThread: public Thread
     CMPServer *parent;
     int mpSoMaxConn;
+    Owned<IThreadPool> threadPool;
+    unsigned acceptThreadPoolSize = 0;
```
will move.

jakesmith commented on system/mp/mpcomm.cpp:
```diff
@@ -448,11 +448,85 @@ class CMPConnectThread: public Thread
+        {
+            CriticalBlock b(crit);
+            for (auto &sock : slowClientsSocks)
+                sock->close();
```
reading up, I think it is unsafe, also, the handling of getting slowClientsSocks.back() in crit, then later calling pop_back() is unsafe is new items are being added.

I am going to remove the socket from the slowClientsSocks before calling handleAcceptedSocket.
It will mean it can't be interrupted and will timeout, but that's not really a problem, just a theoretical slightly longer delay during shutdown if happens to be on a stalling socket.

jakesmith commented:
@ghalliday - please see new commit.

ghalliday reviewed (approved):
@jakesmith please squash

jakesmith commented:
> @jakesmith please squash

@ghalliday - done
